File: rust/src/domain/manga.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Manga {
    pub id: i64,
    pub source: i64,
    pub url: String,
    pub title: String,
    pub artist: Option<String>,
    pub author: Option<String>,
    pub description: Option<String>,
    pub genre: Option<Vec<String>>,
    pub status: MangaStatus,
    pub thumbnail_url: Option<String>,
    pub favorite: bool,
    pub last_update: Option<i64>,
    pub date_added: Option<i64>,
    pub viewer_flags: i32,
    pub chapter_flags: i32,
    pub cover_last_modified: i64,
    pub update_strategy: bool,
    pub initialized: bool,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
pub enum MangaStatus {
    Unknown = 0,
    Ongoing = 1,
    Completed = 2,
    Licensed = 3,
    PublishingFinished = 4,
    Cancelled = 5,
    OnHiatus = 6,
}

impl Default for MangaStatus {
    fn default() -> Self {
        Self::Unknown
    }
}

File: lib/src/domain/manga/model/manga.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'manga.freezed.dart';
part 'manga.g.dart';

@freezed
sealed class Manga with _$Manga {
  const factory Manga({
    required int id,
    required int source,
    required String url,
    required String title,
    String? artist,
    String? author,
    String? description,
    List<String>? genre,
    @Default(MangaStatus.unknown) MangaStatus status,
    String? thumbnailUrl,
    @Default(false) bool favorite,
    DateTime? lastUpdate,
    DateTime? dateAdded,
    @Default(0) int viewerFlags,
    @Default(0) int chapterFlags,
    @Default(0) int coverLastModified,
    @Default(false) bool updateStrategy,
    @Default(false) bool initialized,
  }) = _Manga;

  factory Manga.fromJson(Map<String, dynamic> json) => _$MangaFromJson(json);
}

enum MangaStatus {
  unknown,
  ongoing,
  completed,
  licensed,
  publishingFinished,
  cancelled,
  onHiatus,
}
